src/hotspot/share/oops/cpCache.cpp

```diff
@@ -940,30 +638,41 @@ oop ConstantPoolCache::set_dynamic_call(const CallInfo &call_info, int index) {
   return appendix();
 }
 
+oop ConstantPoolCache::appendix_if_resolved(int method_index) const {
+  ResolvedMethodEntry* method_entry = resolved_method_entry_at(method_index);
+  if (!method_entry->has_appendix())
```

Comment:
If you move the rest of the code in this method into the a new method `ResolvedMethodEntry::appendix_if_resolved()` then you can call that method from here and also call it in places where you have already looked up the `ResolvedMethodEntry` but are still indirecting through the cache method using an `index`.

Reply:
ResolvedMethodEntry doesn't know about the constant pool or the cache, so that mean it will be unable to call `constant_pool()->resolved_reference_at(ref_index);`. Maybe an alternate solution would be to overload `appendix_if_resolved` so that it can take either `int method_index` or `ResolvedMethodEntry* method_entry` as arguments.

Reply:
Doh! Yes, of course. Overloading will allow the repeated lookup to be avoided.